idaes/core/util/model_diagnostics.py

```diff
@@ -17,29 +17,987 @@
 
 __author__ = "Alexander Dowling, Douglas Allan, Andrew Lee"
 
-
 from operator import itemgetter
+from sys import stdout
 
-import pyomo.environ as pyo
-from pyomo.core.expr.visitor import identify_variables
-from pyomo.contrib.pynumero.interfaces.pyomo_nlp import PyomoNLP
-from pyomo.contrib.pynumero.asl import AmplInterface
-from pyomo.core.base.block import _BlockData
 import numpy as np
 from scipy.linalg import svd
 from scipy.sparse.linalg import svds, norm
 from scipy.sparse import issparse, find
 
+from pyomo.environ import (
+    Binary,
+    Block,
+    check_optimal_termination,
+    ConcreteModel,
+    Constraint,
+    Objective,
+    Param,
+    Set,
+    SolverFactory,
+    value,
+    Var,
+)
+from pyomo.core.base.block import _BlockData
+from pyomo.common.collections import ComponentSet
+from pyomo.common.config import ConfigDict, ConfigValue, document_kwargs_from_configdict
+from pyomo.util.check_units import identify_inconsistent_units
+from pyomo.contrib.incidence_analysis import IncidenceGraphInterface
+from pyomo.core.expr.visitor import identify_variables
+from pyomo.contrib.pynumero.interfaces.pyomo_nlp import PyomoNLP
+from pyomo.contrib.pynumero.asl import AmplInterface
+from pyomo.common.deprecation import deprecation_warning
+
 from idaes.core.util.model_statistics import (
+    activated_blocks_set,
+    deactivated_blocks_set,
+    activated_equalities_set,
+    deactivated_equalities_set,
+    activated_inequalities_set,
+    deactivated_inequalities_set,
+    activated_objectives_set,
+    deactivated_objectives_set,
+    variables_in_activated_constraints_set,
+    variables_not_in_activated_constraints_set,
+    degrees_of_freedom,
     large_residuals_set,
     variables_near_bounds_set,
 )
-import idaes.core.util.scaling as iscale
+from idaes.core.util.scaling import (
+    get_jacobian,
+    extreme_jacobian_columns,
+    extreme_jacobian_rows,
+    extreme_jacobian_entries,
+    jacobian_cond,
+)
 import idaes.logger as idaeslog
 
 _log = idaeslog.getLogger(__name__)
 
 
+MAX_STR_LENGTH = 84
+TAB = " " * 4
+
+# TODO: Add suggested steps to cautions - how?
+
+CONFIG = ConfigDict()
+CONFIG.declare("model", ConfigValue(description="Pyomo model object to be diagnosed."))
+CONFIG.declare(
+    "absolute_tolerance",
+    ConfigValue(
+        default=1e-4,
+        domain=float,
+        description="Absolute tolerance to for variables close to bounds.",
+    ),
+)
+CONFIG.declare(
+    "relative_tolerance",
+    ConfigValue(
+        default=1e-4,
+        domain=float,
+        description="Relative tolerance to for variables close to bounds.",
+    ),
+)
```

Comment:
These should have more descriptive names, so we know they only apply to variable-bound differences.